### PR TITLE
change FunctionalComponentOptions to a generic interface

### PR DESCRIPTION
**Pull request by kaorun343**
so as to specify `props` and `slots()`

### CODE REVIEW

**Comment by HerringtonDarkholme** [comment]
Thanks! My concern is typing update is a breaking change for TS users. (I've been bitten by this before). 

Should vue's semver also take this into consideration? Ping @ktsn @yyx990803

**Comment by ktsn** [comment]
hm, as the typings is shipped with core, it should also follow the semver?

**Comment by ktsn** [comment]
To avoid breaking change, how about rename the generic interface and create a specific interface that compatible current type? like:

```js
export interface RenderContextG<P, S> {
  props: P;
  children: VNode[];
  slots: S;
  data: VNodeData;
  parent: Vue;
}

export interface RenderContext extends RenderContext<any, any> {}
```

File: types/options.d.ts [hunk 1]
```diff
@@ -45,17 +45,17 @@ export interface ComponentOptions<V extends Vue> {
...
   name?: string;
 }
 
-export interface RenderContext {
-  props: any;
+export interface RenderContext<P, S> {
+  props: P;
   children: VNode[];
-  slots: any;
+  slots(): S;
```
Comment (ktsn) [inline types/options.d.ts]: Why is this changed to method? In my understand, it should a property of an object.

Comment (kaorun343) [inline types/options.d.ts]: According to a document, it is a function that returns a dictionary.
http://vuejs.org/guide/render-function.html#Functional-Components

**Comment by kaorun343** [comment]
It was a huge breaking change the changes of `ComponentOptions` that I did. However, compared to `ComponentOptions`, it will be a smaller impact to change `FunctionalComponentOptions` and `RenderContext`, I think...

**Comment by ktsn** [comment]
I think `ComponentOptions` update was reasonable breaking change because it was not announced officially about typings and needed revisions for blush up as the stage of pre-release.
However, it is now already released and announced, and it seems many users depends on it.
It may be different problem between code and typings but I think we should be careful about this topic.

**Comment by kaorun343** [comment]
Then, how about changing only `slot: any` -> `slot(): any`?
I'm gonna make pull request again when preparing 2.1...

**Comment by ktsn** [comment]
It's ok to me.
